# 프로미스

> ES6에서는 비동기 처리를 위해 프로미스를 도입 <br> 비동기 처리 시점을 명확하게 표현할 수 있다는 장점

## 비동기 처리를 위한 콜백 패턴의 단점

> 콜백 헬

```js
const get = (url) => {
  const xhr = new XMLHttpRequest();
  xhr.open('GET', url);
  xhr.send();

  xhr.onload = () => {
    if (xhr.status === 200) {
      console.log(JSON.parse(xhr.response));
    } else {
      console.error(`${xhr.status} ${xhr.statusText}`);
    }
  };
};

get('URL');

/**
 * {
 *  "userId": 1,
 *  "id": 1,
 *  "title": "sunt aut facere ...",
 *  "body" : "quia et suscipit ...",
 * }
 */
```

- get 함수는 비동기 함수이다.
  - 비동기 함수를 호출하면 함수 내부의 비동기로 동작하는 코드가 완료되지 않았다 해도 기다리지 않고 즉시 종료된다.
    - 즉, 비동기 함수 내부의 비동기로 동작하는 코드는 비동기 함수가 종료된 이후에 완료
  - 비동기 함수 내부의 비동기로 동작하는 코드에서 처리 결과를 외부로 반환하거나 상위 스코프의 변수에 할당하면 기대한 대로 동작하지 않는다.

```js
let g = 0;

setTimeout(() => {
  g = 100;
}, 0);
console.log(g); // 0
```

- 비동기 함수가 호출되면 함수 코드를 평가하는 과정에서 함수의 실행 컨텍스트가 생성되고 실행 컨텍스트 스택(콜 스택)에 푸시된다.
- 비동기 함수가 종료하면 함수 실행 컨텍스트가 콜스택에서 팝된다.
- 이벤트 핸들러는 이벤트가 발생하면 일단 태스크 큐에 저장되어 대기하다가, 콜 스택이 비면 이벤트 루프에 의해 콜 스택으로 푸시되어 실행된다.

- `callback hell: 콜백 함수 호출이 중첩되어 복잡도가 높아지는 현상`

> 에러 처리의 한계

```js
try {
  setTimeout(() => {
    throw new Error('Error!');
  }, 1000);
} catch (e) {
  console.error('캐치한 에러', e);
}
```

- `setTimeout이` 호출되면 함수의 실행 컨텍스트가 생성되어 콜 스택에 푸시되어 실행되는데 비동기 함수이므로 즉시 종료되어 콜 스택에서 제거된다.
- 이후 타이머가 만료되면 콜백함수는 태스크 큐로 푸시되고 콜 스택이 비었을 때 이벤트 루프에 의해 콜 스택으로 푸시되어 실행된다.
- 콜백 함수가 실행될 때 `setTimeout` 함수는 이미 콜 스택에서 제거되었으므로 콜백 함수의 호출자가 `setTimeout` 함수가 아니다. 에러는 호출자 방향으로 전파되므로 catch에서 잡히지 않는다.

- 에러는 호출자방향으로 전파된다.
  - 즉, 콜 스택의 아래방향으로 전파

## 프로미스의 생성

```js
const promise = new Promise((resolve, reject) => {
    if(/* 비동기 처리 성공 */){
        resolve('result')
    } else {
        reject('failure reason')
    }
})
```

- Promise는 비동기 처리의 진행 상태 정보를 갖는다.

| 프로미스의 상태 정보 | 의미                                  | 상태 변경 조건                   |
| -------------------- | ------------------------------------- | -------------------------------- |
| pending              | 비동기 처리가 아직 수행되지 않은 상태 | 프로미스가 생성된 직후 기본 상태 |
| fulfilled            | 비동기 처리가 수행된 상태(성공)       | resolve 함수 호출                |
| rejected             | 비동기 처리가 수행된 상태(실패)       | reject 함수 호출                 |

## 프로미스의 후속 처리 메서드

- `프로미스의 비동기 처리 상태가 변화하면 후속 처리 메서드에 인수로 전달한 콜백 함수가 선택적으로 호출된다.`

> Promise.prototype.then

- then 메서드는 두 개의 콜백 함수를 인수로 전달받는다.
  - 첫 번째 콜백 함수는 프로미스가 fulfilled 상태가 되면 호출
    - 이때 콜백 함수는 프로미스의 비동기 처리 결과를 인수로 전달 받음
  - 두 번째 콜백 함수는 프로미스가 rejected 상태가 되면 호출
    - 이때 콜백 함수는 프로미스의 에러를 인수로 전달 받음

> Promise.prototype.catch

- catch 메서드는 한 개의 콜백 함수를 인수로 전달받음
  - 프로미스가 rejected 상태인 경우만 호출

> Promise.prototype.finally

- finally 메서드는 한 개의 콜백 함수를 인수로 전달받음

  - finally 메서드의 콜백 함수는 성공 또는 실패와 상관없이 무조건 한 번 호출

- 프로미스 상태와 상관없이 공통적으로 수행해야할 처리 내용이 있을 때 유용

## 프로미스의 에러 처리

- 프로미스는 에러를 문제없이 처리 가능

  - then, catch, finally를 사용하여 수행

- catch 메서드를 모든 then 메서드를 호출한 뒤 호출하면 비동기 처리에서 발생한 에러와 then 메서드 내부에서 발생한 에러까지 모두 캐치 가능
  - 에러처리는 catch 메서드에서 하는 것을 권장

## 프로미스 체이닝

```js
const url = "url";

promiseGet(`${url}/posts/1`)
  .then(({userId} => promiseGet(`${url}/users/${userId}`)))
  .then(userInfo => console.log(userInfo))
  .catch(err => console.log(err));
```

- then -> then -> catch 순서로 후속 메서드를 호출

  - 후속 처리 메서드는 언제나 프로미스를 반환하므로 연속적으로 호출 가능
  - 프로미스 체이닝

- 문제점

  - 콜백 패턴은 가독성이 좋지 않다.
  - 콜백 헬에 빠짐

- async/await를 통해 해결 가능
  - 프로미스의 후속 처리 메서드 없이 마치 동기 처리처럼 프로미스가 처리 결과를 반환하도록 구현 가능

## 프로미스의 정적 메서드

> Promise.resolve/Promise.reject

- Promise.resolve

  - 인수로 전달받은 값을 resolve하는 프로미스 생성

- Promise.reject
  - 인수로 전달받은 값을 reject하는 프로미스 생성

> Promise.all

- 여러 개의 비동기 처리를 모두 병렬처리할 때 사용

```js
const requestData1 = () => {
  new Promise((resolve) => setTimeout(() => resolve(1), 3000));
};
const requestData2 = () => {
  new Promise((resolve) => setTimeout(() => resolve(2), 2000));
};
const requestData3 = () => {
  new Promise((resolve) => setTimeout(() => resolve(3), 1000));
};

Promise.all([requestData1(), requestData2(), requestData3()])
  .then(console.log) // [1, 2, 3] 약 3초 소요
  .catch(console.error);
```

> Promise.race

- Promise.all 메서드와 동일하게 프로미스를 요소로 갖는 배열 등의 이터러블을 인수로 전달받는다.
- 가장 먼저 fulfilled 상태가 된 프로미스의 처리 결과를 resolve하는 새로운 프로미스를 반환

```js
Promise.race([
  new Promise((resolve) => setTimeout(() => resolve(1), 3000)),
  new Promise((resolve) => setTimeout(() => resolve(2), 2000)),
  new Promise((resolve) => setTimeout(() => resolve(3), 1000)),
])
  .then(console.log) // 3
  .catch(console.error);
```

> Promise.allSettled

- Promise.allSettled 메서드는 프로미스를 요소로 갖는 배열 등의 이터러블을 인수로 전달받는다.
- settled 상태가 되면 처리 결과를 배열로 반환

## 마이크로태스크 큐

```js
setTimeout(() => console.log(1), 0);

Promise.resolve(())
  .then(() => console.log(2))
  .then(() => console.log(3));
```

- 1 -> 2 -> 3의 순으로 출력될 것처럼 보이지만 2 -> 3 -> 1로 출력됨
- 프로미스의 후속 처리 메서드의 콜백 함수는 태스크 큐가 아니라 마이크로태스크 큐에 저장되기 때문

- 마이크로 태스크 큐는 태스크 큐보다 우선순위가 높다.

## fetch

- XMLHttpRequest 객체와 마찬가지로 HTTP 요청 전송 기능을 제공하는 클라이언트 사이드 Web API이다.

- fetch 함수가 반환하는 프로미스는 기본적으로 `404 Not Found나 500 Internal Server Error와 같은 HTTP 에러`가 발생해도 에러를 reject하지 않고 불리언 타입의 ok 상태를 false로 설정한 `Response 객체를 resolve`한다.
- `오프라인 등의 네트워크 장애나 CORS 에러에 의해 요청이 완료되지 못한 경우에만 reject한다.`

- 반면에 axios는 모든 HTTP 에러를 reject하는 프로미스를 반환
